Run step for soca (grid generation and 3dvar fgat)

guillaumevernieres opened the pull request:
## Description

This PR adds the necessary features to allow testing of:
- The `soca` grid generation 
- The generation of the flow dependent parametric diagonal of B
- 3DVARFGAT with a correlation operator set to Id

All of the above are executed within [exgdas_global_marine_analysis_run.sh](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_run.sh). 
The general idea is to attempt to replicate the environment that will be inherited from the global workflow and run the `prep` and `run` scripts with toy backgrounds within `ctests`. I don't expect the `prep` or `run` steps to work as is within the global workflow. That specific work will be the subject of an other PR.

## Features added:
- The [exgdas_global_marine_analysis_prep.py](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_prep.py) scripts does the following:
  - Creates the directories to run stuff in and dump outputs into
  - prepares all the static files
  - fetch the obs
  - Link the [gridgen.yaml](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/parm/soca/gridgen/gridgen.yaml) used to run the grid generation application
  - Generates the 3DVAR FGAT `yaml` files, linking to the 3 hourly MOM6 backgrounds
- The [exgdas_global_marine_analysis_run.sh](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_run.sh) runs `soca_gridgen.x` and `soca_var.x`
- A ctest for `exgdas_global_marine_analysis_run.sh`:
  - It is supposedly mimicking what would happen when ran with the global workflow. But it's currently probably quite naive, and maybe some of it is wrong ... The mostly guessed environment is in [runtime_vars.sh](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/test/soca/runtime_vars.sh)
  - It is staging all the restarts and required static files from the `soca` repo, which uses a very low-res test case (5deg)

## Testing
The new features have been tested on `Hera`, 'Orion`, my personal laptop and desktop.

 ## Issue addressed
- fixes #29

guillaumevernieres committed:
...

guillaumevernieres committed:
ctest for marine run. Does not work.

guillaumevernieres committed:
Merge branch 'develop' into feature/soca-run

guillaumevernieres committed:
gridgen working

guillaumevernieres committed:
added missing files

guillaumevernieres committed:
3dvarfgat working

guillaumevernieres committed:
Merge branch 'develop' into feature/soca-run

guillaumevernieres committed:
fixed code norms, removed hardcoded container

guillaumevernieres committed:
make run step work on Hera

guillaumevernieres commented:
@CoryMartin-NOAA , @RussTreadon-NOAA , it is still a draft PR, but I'm marking as ready to discuss. 
I was not able to make it work on `Orion`:
- I'm having weird issues with python
- The `run_step` `ctests` is failing in the `ci`. Although the grid generation runs to completion but it is failing when checking for the return value (?):
```console
  terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
srun: error: Orion-02-62: tasks 0-1: Aborted
srun: launch/slurm: _step_signal: Terminating StepId=5723194.0
```
but it does terminate with a `status = 0`:
```console
# See here on Orion:
/work2/noaa/stmp/cmartin/CI/GDASApp/PR/62/GDASApp/build/test/soca/3dvar/analysis/exgdas_global_marine_analysis_run.log

[...]
OOPS_STATS ---------------------------------- Parallel Timing Statistics (   2 MPI tasks) -----------------------------------

OOPS_STATS Run end                                  - Runtime:      4.88 sec,  Memory: total:   529.69 Mb, per task: min =   259.68 Mb, max =   270.01 Mb
Run: Finishing soca::GridGen< with status = 0
```

I'll look it this issue more closely tomorrow.

guillaumevernieres committed:
test for errors outside of run/prep scripts

guillaumevernieres committed:
...

guillaumevernieres committed:
removed debugging prints

guillaumevernieres committed:
code norm

guillaumevernieres committed:
moved D from B outside of the var changes

guillaumevernieres committed:
Merge branch 'feature/soca-run' of https://github.com/NOAA-EMC/GDASApp into feature/soca-run

guillaumevernieres committed:
Merge branch 'develop' into feature/soca-run

guillaumevernieres committed:
code norm/tidy

guillaumevernieres committed:
Merge branch 'develop' into feature/soca-run

guillaumevernieres committed:
Merge branch 'feature/soca-run' of https://github.com/NOAA-EMC/GDASApp into feature/soca-run

FanHan-NOAA reviewed (dismissed):
Checkout the branch and all ctests passed on Orion. Code changes look good to me, although I'm not an expert on soca. Just out of curiosity, is it possible to run a standalone marine analysis with `run_jedi_exe.py`?

guillaumevernieres commented:
> Checkout the branch and all ctests passed on Orion. Code changes look good to me, although I'm not an expert on soca. Just out of curiosity, is it possible to run a standalone marine analysis with `run_jedi_exe.py`?

Thanks for your review @FanHan-NOAA . 
It isn't possible (yet?). But, the ctests that I added do run 1 cycle of a marine analysis at very low res (5deg). One ctest runs the [marine prep script](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_prep.py) and the subsequent one runs the [marine run script](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_run.sh) (where the JEDI DA is called).
This is probably how `run_jedi_exe.py` should be refactored, calls to the prep and run steps, and probably added to our ctests.

CoryMartin-NOAA commented on parm/soca/obs/config/adt_j3.yaml:
```diff
@@ -1,9 +1,9 @@
   name: adt_j3
   obsdatain:
-    obsfile: ./obs/adt_j3.{{window_begin}}.nc4
+    obsfile: ./obs/adt_j3.nc4
```
We should conform to the same file naming convention here, right?

CoryMartin-NOAA commented on parm/soca/obs/config/adt_j3.yaml:
```diff
@@ -1,9 +1,9 @@
+    obsfile: ./obs/adt_j3.nc4
   obsdataout:
-    obsfile: ./obs_out/adt_j3.{{window_begin}}.nc4
+    obsfile: ./obs_out/adt_j3.$(OBS_DATE).nc4
```
same comment, this should probably be consistent across components?

CoryMartin-NOAA commented on parm/soca/obs/config/sst_noaa19_l3u.yaml:
```diff
@@ -1,9 +1,9 @@
+    obsfile: ./obs/sst_noaa19_l3u.nc4
   obsdataout:
-    obsfile: ./obs_out/sst_noaa19_l3u.{{window_begin}}.nc4
+    obsfile: ./obs_out/sst_noaa19_l3u.$(OBS_DATE).nc4
```
Same comment for these two lines as above.

CoryMartin-NOAA commented on scripts/exgdas_global_marine_analysis_prep.py:
```diff
@@ -59,16 +59,23 @@ def gen_bkg_list(bkg_path='.', file_type='MOM', yaml_name='bkg.yaml'):
 ufsda.mkdir(anl_dir)
 
+# create output directory for obs
+ufsda.mkdir(os.path.join(comout, 'analysis', 'obs_out'))
```
"diags"

CoryMartin-NOAA reviewed (commented):
Couple of comments but mostly looks good to me.

guillaumevernieres committed:
conform to naming convention

guillaumevernieres committed:
fixed run step

guillaumevernieres committed:
Merge branch 'develop' into feature/soca-run

guillaumevernieres commented:
@CoryMartin-NOAA , I addressed your comments but the CI is failing. I'll look into it today again, but the last commit I push was testing fine on `Orion`.

CoryMartin-NOAA commented:
@guillaumevernieres ok, it could also be a CI issue. I can take a look later today as well.

guillaumevernieres commented:
> @guillaumevernieres ok, it could also be a CI issue. I can take a look later today as well.

@CoryMartin-NOAA ... `soca` develop is failing now. We should probably do something similar to what @travissluka did for `soca` and `soca-science` and have the latest working develop saved in a daily release branch ... one day ...

FanHan-NOAA reviewed (approved):
Cloned a new GDASApp, built with fv3-bundle, and all ctests passed for this branch on Orion. Changes look good to me, thanks!
